Make query termination more deterministic

systay opened the pull request.

Commit:
Make query termination more deterministic

pontusmelke commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/KillQueryTest.scala:
```diff
@@ -0,0 +1,131 @@
+  val THREAD_COUNT = 10
+
+
+  test("apa") {
```
a more descriptive name maybe :)

pontusmelke commented on community/kernel/src/main/java/org/neo4j/kernel/impl/query/Neo4jTransactionalContext.java:
```diff
@@ -56,6 +59,7 @@
     private InternalTransaction transaction;
     private Statement statement;
     private boolean isOpen = true;
+    private boolean terminated = false;
```
AtomicBoolean?

Commit:
Minor tweaks

lutovich committed:
Make Neo4jTransactionalContext termination safe

 * ask the underlying transaction about it's termination status instead
   of having a separate termination flag
 * make transaction field volatile because it can be accessed from a
   different thread during termination
 * allow closing after termination because it is still required to
   free all kernel transaction resources

lutovich committed:
Add ITs for RESET in bolt

New class `SessionResetIT` contains tests that exercise RESET
functionality via `Session#reset()`. Also added a comment explaining
transaction termination in bolt state machine.

MishaDemianenko commented on community/bolt/src/main/java/org/neo4j/bolt/v1/runtime/TransactionStateMachine.java:
```diff
@@ -95,6 +95,14 @@ public void streamResult( ThrowingConsumer<BoltResult, Exception> resultConsumer
+     * Rollback and close transaction. Move back to {@link State#AUTO_COMMIT}.
+     * <p>
+     * <b>Warning:</b>This method should only be called by the bolt worker thread during it's regular message
+     * processing. It is wrong to call it from a different thread because kernel transactions are not thread-safe.
```
Can we in some way force that to be a runtime check?

lutovich replied:
We can propagate bolt worker thread id from `RunnableBoltWorker` to `BoltStateMachine` and from there to this `TransactionStateMachine`. Then we can probably add runtime checks like this in all methods:

```java
if(Thread.currentThread() != boltWorkerThread)
    throw new IllegalStateException();
```

Do you think it is worth doing? Maybe as a separate PR?

MishaDemianenko replied:
I think by doing that we will guard us against misuse since the comment is something that is not enforced; for sure that can be separate PR.

MishaDemianenko commented on integrationtests/src/test/java/org/neo4j/bolt/SessionResetIT.java:
```diff
@@ -0,0 +1,361 @@
+    {
+        AtomicReference<Session> sessionRef = new AtomicReference<>();
+
+        Future<Void> queryResult = runAsync( () ->
```
Since this is CompletableFuture, we can avoid sleep and awaits here completely and just reset the session when query result future will be completed right?

lutovich replied:
Future here executes query and we want to interrupt query execution with `#reset()` so can't wait for completion. Waiting for result happens only when asserting on success/failure. Subsequent `await` wants to see that query execution is in progress.

MishaDemianenko replied:
ok

MishaDemianenko reviewed (commented):
Couple of small comments otherwise looks fine!

benbc commented:
@systay @MishaDemianenko Is there not a fix for this that could go into a 3.1 point release?

MishaDemianenko commented:
@benbc it can go into 3.1 but in the same time neither me no @lutovich have seen those problems in 3.1 logs so we would like to avoid touching 3.1 for now

benbc commented:
@MishaDemianenko The original problem that I reported which spawned this work was in 3.1.

MishaDemianenko commented:
@benbc yes I know - benchmarks store; but since that is not really customer reported issue we would like to avoid touching 3.1 since we do not feel comfortable to introduce that change there so far.
